src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -119,42 +128,241 @@ private LabelSymbol GetDagNodeLabel(BoundDecisionDag dag)
             /// </summary>
             private void LowerDecisionDag(ImmutableArray<BoundDecisionDag> sortedNodes)
             {
+                switch (sortedNodes[0])
+                {
+                    case BoundWhenClause wc:
+                        // If the first node is in a when clause's section rather than the code for the
+                        // lowered decision dag, jump there to start.
+                        _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(wc)));
+                        break;
+                    case BoundDecision d:
+                        // If the first node is a decision rather than the code for the
+                        // lowered decision dag, jump there to start.
+                        _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(d)));
+                        break;
+                }
+
                 // Call LowerDecisionDagNode with each node and its following node in the generation order.
+                // However, some nodes can be emitted more efficiently as a switch dispatch (possibly out
+                // of order).
+                var loweredNodes = PooledHashSet<BoundDecisionDag>.GetInstance();
                 BoundDecisionDag previous = null;
                 foreach (BoundDecisionDag node in sortedNodes)
                 {
+                    if (node is BoundWhenClause w)
+                    {
+                        // Code for the when clause goes in a separate code section for the switch section.
+                        LowerWhenClause(w);
```

Comment:
Does this expect the previous node to have been lowered already? If so, we should probably assert that fact.

Reply:
No; this just removes any `BoundWhenClause` from being included in the `previous`/`node` sequence, since the code will not be mixed with the other code of the decision tree. The `continue` goes to the next iteration of the loop for `previous` to be dealt with.

This code pattern of dealing with previous/next is much simplified in a later PR, where we create a new list of nodes to lower that excludes them.

---
In reply to: [177229494](https://github.com/dotnet/roslyn/pull/25544#discussion_r177229494) [](ancestors = 177229494)